pyro/nn/auto_reg_nn.py

```diff
@@ -7,6 +7,75 @@
 from torch.nn import functional as F
 
 
+def sample_mask_indices(input_dim, hidden_dim, simple=False, conditional=False):
+    """
+    Samples the indices assigned to hidden units during the construction of MADE masks
+
+    :param input_dim: the dimensionality of the input variable
+    :type input_dim: int
+    :param hidden_dim: the dimensionality of the hidden layer
+    :type hidden_dim: int
+    :param simple: True to sample indices uniformly, false to space indices evenly and round up or down randomly
+    :type simple: bool
+    :param conditional: True if sampling indices for a conditional-MADE, false otherwise
+    :type conditional: bool
+    """
+    start_integer = 0 if conditional else 1
+
+    if simple:
+        # Simple procedure samples each index i.i.d.
+        return torch.randint(start_integer, input_dim, size=(hidden_dim,))
+    else:
+        # "Non-simple" procedure creates fractional indices evenly then rounds at random
+        mk = torch.linspace(start_integer, input_dim - 1, steps=hidden_dim)
+
+        # Randomly decide whether to round up or down indices
+        ints = mk.floor()
+        ints += torch.bernoulli(mk - ints)
+        return ints
+
+
+def create_mask(input_dim, observed_dim, hidden_dim, num_layers, permutation, output_dim_multiplier):
+    """
+    Creates MADE masks for a conditional distribution
+
+    :param input_dim: the dimensionality of the input variable
+    :type input_dim: int
+    :param observed_dim: the dimensionality of the variable that is conditioned on (for conditional densities)
```

Comment:
as i understand it this `observed_dim` functionality isn't being tested anywhere. can we add a test for this?

Reply:
Not at all! It's my first open-source contribution and a fun learning experience. :)

I think I've run it on the GPU, but let me double check.

Yes for the observed_dim, I can add a test that checks that *all* members of the jacobian are effected when you change the conditioning variable

Reply:
I've removed observed_dim as I realized it isn't needed